packages/villus/src/utils/common.ts

```diff
@@ -16,13 +17,17 @@ export function toWatchableSource<T = any>(value: Ref<T> | Record<string, any>):
 
 // Uses same component provide as its own injections
 // Due to changes in https://github.com/vuejs/vue-next/pull/2424
+// todo: maybe better to modify func name and add manual client args here
 export function injectWithSelf<T>(symbol: InjectionKey<T>, onMissing: () => Error): T {
   const vm = getCurrentInstance() as any;
+  let client = vm && inject(symbol, vm?.provides?.[symbol as any]);
 
-  const injection = inject(symbol, vm?.provides?.[symbol as any]);
-  if (injection === null || injection === undefined) {
+  if (client) setActiveClient(client);
```

Comment:
This could be dangerous to set here. Assuming either `useClient` or `createClient` were called anywhere previously this would make this redundant, right? Unless you are mitigating something I'm not aware of.

Reply:
Yes, I do have this worry at first, I think if we just create one client, it is not a problem. If we can create multi Client, due to the setActiveClient comes form current instance(If we can get vm), so setting it to default client is not a problem(I think), if who want use custom client (diff from vm), they should set a manualClient param ( this is why I want add a manual param to extend multi client.)
Maybe it's ok for one client, but you are right, If we use multi client, the order of setActiveClient anywhere may cause some error, I need some time to check it, then update this later.

Reply:
I have some info, due to the solution comes from pinia, I found that we only create pinia instance once, then defineStore and then useTheDefinedStore. But when we use multi villus client, we may create multi times. and the useQuery function is not depends on client. If the setActiveClient order is not managed, it may get unexpected response. 
I think when we use multi villusClient, It should call useQuery to offer a specific client ( the manualClient param.), or we can suggest user to define customer useQuery in docs, just like (it is an easy and more reliable ):

```ts
export function useCustomClientQuery (opt: QueryOptions) {
    return useQuery(opt, customClient)
}
```

or villus can also have define function to extend more abilities. 

by the way, I think when we can get client from current instance, set it to default and use it to query is reasonable. Developers should aware of which client provided in current instance. if not , pass a specific client is reliable (recommended). I think we can make it more clear in docs.

Above all is my simple opinion. do you have any other good idea?

Reply:
So what I'm thinking is, if we set the active client there it will mess up when the user uses `useQuery` or `useMutation` when it is both outside `setup` and they have multiple clients. It will only work for them if they expect the last created client to be injected.

If we don't do that, the same problem still occurs. Its just we only set it once and that's it. If they use a different client and expect that to be used outside the `setup` it will mess up then.

So regardless of our choices here this would cause issues for people using multiple clients. 
I'm willing to argue that using multiple clients at the same app isn't common and usually people stitch schemas on the backend, so this might be fine to leave as is.

Like you said the `manualClient` option would solve it for those users in that case, which can be documented to make sure they don't fall into that trap.

Reply:
> when it is both outside setup and they have multiple clients. It will only work for them if they expect the last created client to be injected.

It will use last currentInstance used client to execute query,  because we may set activeClient in last currentInstance (resolveClient function). If we don't set, villus will execute with last created client. both will have some risk (if use multi client). I think last currentInstance client maybe more probability which user want choose. 

> So regardless of our choices here this would cause issues for people using multiple clients.

I do have the worry. but I have no more good idea now. so let's add it todo. and make a caution in documentation first to  remind multi client user.  Maybe someday new idea will come.

Reply:
By the way, because we will call createClient in useClient function. if we useClient before query, the activeClient will be set to the right one. So I think some one can useClient first as before (Maybe).